File: CRM/src/main/java/com/sgdcbrk/crm/business/abstracts/OpportunityService.java
package com.sgdcbrk.crm.business.abstracts;

import com.sgdcbrk.crm.dto.opportunity.request.OpportunityRequest;
import com.sgdcbrk.crm.dto.stats.ChartData;
import com.sgdcbrk.crm.model.opportunity.Opportunity;

import java.util.List;
import java.util.Map;

public interface OpportunityService {
    void saveOpportunity(OpportunityRequest request);
    void updateOpportunity(long id, OpportunityRequest request);
    void deleteOpportunityById(long id);

    // Change status
    void completeOpportunityWithWon(Opportunity opportunity);
    void completeOpportunityWithLost(Opportunity opportunity);
    // queries - value
    List<Opportunity> getOpportunitiesByIncreaseValue();
    List<Opportunity> getOpportunitiesByDecreaseValue();

    // queries - Date
    List<Opportunity> getOpportunitiesByShortestDate();
    List<Opportunity> getOpportunitiesByLongestDate();

    // search
    List<Opportunity> searchOpportunities(String searchTerm);

    // default get functions
    Opportunity getOpportunity(long id);
    List<Opportunity> getOpportunities();

    // stats - for charts
    List<ChartData> getOpportunitiesByCompany();
    long getOpenOpportunitiesCount();
    long getWonOpportunitiesCount();
    long getLostOpportunitiesCount();
    List<ChartData> getTopCustomersByOpportunities();
}

File: CRM-Angular/src/app/services/Opportunity/opportunity.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { OpportunityRequest } from '../../dto/Opportunity/Opportunity-request';
import { Observable } from 'rxjs';
import { Opportunity } from '../../dto/Opportunity/Opportunity';
import {ConfigService} from "../../config/config.service";

@Injectable({
  providedIn: 'root',
})
export class OpportunityService {
  private readonly apiUrl;

  constructor(private http: HttpClient, private configService:ConfigService) {
    this.apiUrl = this.configService.getApiUrl() + '/opportunities';
  }

  create(request: OpportunityRequest): Observable<any> {
    return this.http.post(`${this.apiUrl}`, request, {
      responseType: 'text' as 'json',
    });
  }

  update(id: number, request: OpportunityRequest): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}`, request, {
      responseType: 'text' as 'json',
    });
  }

  delete(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`, {
      responseType: 'text' as 'json',
    });
  }

  getOpportunity(id: number): Observable<any> {
    return this.http.get(`${this.apiUrl}/${id}`);
  }

  getOpportunities(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}`);
  }

  completeOpportunityWithWon(id: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/${id}/won`, {
      responseType: 'text' as 'json',
    });
  }

  completeOpportunityWithLost(id: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/${id}/lost`, {
      responseType: 'text' as 'json',
    });
  }

  // -- -- -- -- QUERIED LISTS -- -- -- //

  // Value Orders
  getOpportunitiesByValueOrderAsc(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/value-asc`);
  }

  getOpportunitiesByValueOrderDesc(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/value-desc`);
  }

  // Date Orders
  getOpportunitiesByDateOrderAsc(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/date-asc`);
  }

  getOpportunitiesByDateOrderDesc(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/date-desc`);
  }

  // Search
  getOpportunityBySearch(search: string): Observable<any> {
    const params = new HttpParams().set('search', search);

    return this.http.get<any>(`${this.apiUrl}/search`, { params });
  }

  // stats //
  //---------
  // Şirket başına fırsat sayısını getiren metod
  getOpportunitiesByCompany(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/stats/by-company`);
  }

  // Açık fırsatların sayısını getiren metod
  getOpenOpportunitiesCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/stats/open`);
  }

  // Kazanılmış fırsatların sayısını getiren metod
  getWonOpportunitiesCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/stats/won`);
  }

  // Kaybedilmiş fırsatların sayısını getiren metod
  getLostOpportunitiesCount(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/stats/lost`);
  }

  // En fazla fırsat getiren müşteri bilgisini getiren metod
  getTopCustomerByOpportunities(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/stats/top-customer`);
  }
}
